src/VisualStudio/Core/Def/ValueTracking/ValueTrackingToolWindow.cs

```diff
@@ -49,28 +64,33 @@ public ValueTrackingToolWindow() : base(null)
             Content = _root;
         }
 
-        public ValueTrackingToolWindow(ValueTrackingTreeViewModel viewModel)
-            : base(null)
+        public void Initialize(ValueTrackingTreeViewModel viewModel, Workspace workspace, IThreadingContext threadingContext)
         {
-            Caption = ServicesVSResources.Value_Tracking;
-            Content = _root;
+            Contract.ThrowIfTrue(Initialized);
+
+            Initialized = true;
             ViewModel = viewModel;
+            _workspace = workspace;
+            _threadingContext = threadingContext;
+
+            _workspace.WorkspaceChanged += OnWorkspaceChanged;
         }
 
-        public TreeItemViewModel? Root
+        private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
         {
-            get => ViewModel?.Roots.Single();
-            set
-            {
-                if (value is null)
-                {
-                    return;
-                }
+            Contract.ThrowIfFalse(Initialized);
 
-                Contract.ThrowIfNull(ViewModel);
+            if (e.Kind is WorkspaceChangeKind.SolutionCleared
+                or WorkspaceChangeKind.SolutionRemoved)
+            {
+                RoslynDebug.AssertNotNull(_threadingContext);
+                RoslynDebug.AssertNotNull(ViewModel);
 
-                ViewModel.Roots.Clear();
-                ViewModel.Roots.Add(value);
+                _ = _threadingContext.JoinableTaskFactory.RunAsync(async () =>
+                {
+                    await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync();
+                    ViewModel.Roots.Clear();
+                });
```

Comment:
Does this need to run under either the shutdown token or something else? Otherwise my worry is if this runs "later" than you might expect it to.

Reply:
As long as this runs *eventually* it should be fine. I assume that's at least reasonably seconds and not minutes/hours for a user.

Reply:
I guess I wasn't clear: my concern is if it runs too _late_, perhaps if other things have already shut down in ways you didn't expect.

Reply:
The only thing that could cause it to be a problem is if the threadingContext is gone. The toolwindow lifetime is aligned with package lifetime, so I don't think there _can_ be a problem here. If this doesn't ever get called because of package unload that's also fine!

Reply:
As long as there's not a time that this runs so late that this will crash, which can happen sometimes if something it uses (even implicitly) becomes unavailable.